cryodrgn/source.py

```diff
@@ -0,0 +1,400 @@
+        return ImageSourceIterator(self, chunksize=chunksize)
+
+
+class ArraySource(ImageSource):
+    def __init__(self, data: np.ndarray):
+        if data.ndim == 2:
+            data = data[np.newaxis, ...]
+        nz, ny, nx = data.shape
+        assert ny == nx, "Only square arrays supported"
+        self.data = data
+
+        super().__init__(L=ny, n=nz)
+
+    def _images(self, indices: np.ndarray):
+        return self.data[indices, ...]
+
+
+class MRCFileSource(ImageSource):
+    def __init__(self, filepath: str, *args, **kwargs):
+        from cryodrgn.mrc import MRCHeader
+
+        header = MRCHeader.parse(filepath)
+        self.header = header
+        self.mrcfile_path = filepath
+        self.dtype = header.dtype
+        self.start = 1024 + header.fields["next"]  # start of image data
+        self.nz, self.ny, self.nx = (
+            header.fields["nz"],
+            header.fields["ny"],
+            header.fields["nx"],
+        )
+        assert self.ny == self.nx, "Only square images supported"
+        self.size = self.ny * self.nx
+        self.stride = self.dtype().itemsize * self.size
+
+        super().__init__(L=self.ny, n=self.nz, filenames=filepath, *args, **kwargs)
+
+    def _images(
+        self,
+        indices: np.ndarray,
+        data: Optional[np.ndarray] = None,
+        tgt_indices: Optional[np.ndarray] = None,
+    ):
+
+        with open(self.mrcfile_path) as f:
+            if data is None:
+                data = np.empty((len(indices), self.L, self.L), dtype=self.dtype)
+                assert (
+                    tgt_indices is None
+                ), "Target indices can only be specified when passing in a preallocated array"
+                tgt_indices = np.arange(len(indices))
+            else:
+                if tgt_indices is not None:
+                    assert len(tgt_indices) == len(
+                        indices
+                    ), "indices/tgt_indices length mismatch"
+                else:
+                    tgt_indices = np.arange(len(indices))
+
+            assert isinstance(tgt_indices, np.ndarray)
+
+            for (index, tgt_index) in zip(indices, tgt_indices):
+                f.seek(self.start)
+                offset = index * self.stride
+                # 'offset' in the call below is w.r.t the current position of f
+                _data = np.fromfile(
+                    f, dtype=self.dtype, count=self.size, offset=offset
+                ).reshape(self.ny, self.nx)
+                data[tgt_index, : self.ny, : self.nx] = _data
+
+            return data
+
+
+class TxtFileSource(ImageSource):
+    def __init__(self, filepath: str, n_workers: int = 1, *args, **kwargs):
+
+        _paths = []
+        for line in open(filepath).readlines():
+            line = line.strip()
+            if not os.path.isabs(line):
+                _paths.append(os.path.join(os.path.dirname(filepath), line))
+            else:
+                _paths.append(line)
```

Comment:
Up to you, but slightly nicer...

```suggestion
        _paths = []
        reldir = os.path.dirname(filepath)
        for line in open(filepath).readlines():
            path = line.strip()
            if not os.path.isabs(path)
                path = os.path.join(reldir, path)
            _paths.append(path)
```

Reply:
Yes - definitely nicer. I chose `filepath_dir` over `reldir` though. Hope that works. Added in b2f85194f01817c24ca579a66f6014628930d29e